RFC: hook up max spell targets from dbc info

nuoHep opened the pull request:
Not sure if worth. Also, I suck at naming things, not even sure if `max_targets` is how this should be called.

@navv1234 dunno where you are at with your dbc export refactorings, just fyi that we may need this info

nuoHep committed:
[DBC] Hook up spell max targets

navv1234 commented:
Yeah, that's fine and useful info for us in certain situations.

My intent with the refactoring is to split everything into their own files instead of growing our primary structs even more (spell, item). For that to happen we need to figure out a way to do something like what you had (the pseudo-runtime linking of data X with our "spell_data_t" using spans or something similar).

That change will also mean that the whole interface on how we access that data could change, as there might be multiple entries in a linked db2 file for a single spell (might not be true for player-related spells though, this is more on the enemy side).

navv1234 commented on dbc_extract3/dbc/generator.py:
```diff
@@ -2856,6 +2863,11 @@ def generate(self, ids = None):
+            targeting = spell.get_link('targeting')
+            fields += targeting.field('max_affected_targets')
+            f, hfd = targeting.get_hotfix_info(('max_affected_targets', 48))
+
```
You should be able to do something like:
```
targeting = spell.children('SpellTargetRestrictions')
if len(targeting):
    fields += targeting[0].field('max_affected_targets')
    f, hfd = targeting[0].get_hotfix_info(('max_affected_targets', 48))
else:
    fields += self.db('SpellTargetRestrictions').default().field('max_affected_targets')
```

This is kinda clunky though, so I'm also slightly wondering if that children() interface should change to return a default entry if there are no actual children available for the parent. Note also that children() will return a list of entries, because technically, a single parent record can have multiple child records in a db2 file.

nuoHep replied:
That's much nicer, yeah. I decided to go a bit further and introduced a `.child()` accessor. Not sure... but for stuff that we either know there is only 1 thing (or we simply don't care) it seems it's much easier.

navv1234 replied:
Yeah that's fine, although it does hide problems if blizzard makes an unconventional spell data change (adds another child record and the ordering is such that the old one is no longer the first entry).

nuoHep replied:
I don't think we would be able to handle such a change for stuff we expect a 1-to-1 mapping even with an explicit `children()[0]` access. It would also hide the problem from us.

What I think we can do for data we expect a 1-to-1 link is to slap a warning inside `child()` in case the number of linked records is more than one. It would actually better protect us from changes in spell data.

navv1234 replied:
No, of course not, better way would be to analyze the data before generation and warn about multiple entries.

EDIT: And yes, can add the warning to child() also.

navv1234 commented on dbc_extract3/dbc/data.py:
```diff
@@ -643,6 +643,9 @@ def initialize_data_model(options):
 
+        if options.build >= dbc.WowVersion(9, 0, 0, 0):
+            SpellName.link('targeting', SpellTargetRestrictions)
+
```
If you go by the new data model stuff (see comment below where you add the field to spell_data_t record output), linking is completely unnecessary, if the current format file defines this linkage (i.e., the `parent` value for `SpellTargetRestrictions` is set to `SpellName` in the format file).

navv1234 commented on dbc_extract3/dbc/generator.py:
```diff
@@ -1993,6 +1993,9 @@ def __init__(self, options, data_store):
 
+        if self._options.build >= dbc.WowVersion(9, 0, 0, 0):
+            self._dbc.append('SpellTargetRestrictions')
+
```
If you go by the new data model stuff (see comment below where you add the field to spell_data_t record output), explicitly instructing the generator to load db2 files is unnecessary.

navv1234 commented on dbc_extract3/dbc/generator.py:
```diff
@@ -2029,6 +2032,10 @@ def initialize(self):
+
+        if self._options.build >= dbc.WowVersion(9, 0, 0, 0):
+            self._data_store.link('SpellTargetRestrictions', 'id_parent', 'SpellName', 'targeting')
+
```
If you go by the new data model stuff (see comment below where you add the field to spell_data_t record output), explicitly linking stuff into a parent db2 file is unnecessary, **if** the current format file defines this linkage (i.e., the `parent` value for `SpellTargetRestrictions` is set to `SpellName` in the format file).

navv1234 commented on engine/dbc/sc_spell_info.cpp:
```diff
@@ -1435,6 +1435,14 @@ std::string spell_info::to_str( const dbc_t& dbc, const spell_data_t* spell, int
+  if ( spell -> max_targets() != 0 )
+  {
+    fmt::print( s, "Max Targets      : {}{}{}\n",
+      spell -> max_targets() < 0 ? "Unlimited(" : "",
```
Might make sense to call this ´Unknown´ instead of Unlimited, from the spell query dump we can see some values of at least -1 and -6, meaning Blizzard might be using negative values to denote some thing server-sidey.

nuoHep replied:
Yeah, so `-1` is def unlimited from what I've seen. So I decided to do:
* `Unlimited(-1)` if n == -1
* `Unknown(n)` otherwise if n < 0

navv1234 replied:
Yep, that's good.

nuoHep committed:
[DBC] Add RawDBCRecord.child()

For use with databses where we either know there is only a single record or we
don't care and need only the first one.

XXX: Not sure about error handling.

nuoHep committed:
Regenerate spell data

nuoHep committed:
Update spell data dump

nuoHep commented:
> My intent with the refactoring is to split everything into their own files instead of growing our primary structs even more (spell, item). For that to happen we need to figure out a way to do something like what you had (the pseudo-runtime linking of data X with our "spell_data_t" using spans or something similar).
> 
> That change will also mean that the whole interface on how we access that data could change, as there might be multiple entries in a linked db2 file for a single spell (might not be true for player-related spells though, this is more on the enemy side).

I'm not sure about splitting `spell_data_t` tbh. While it's natural for labels, powers and (especially) effects (it's always 1-to-n and the linked data is somewhat big) I'd be very cautious about splitting off other stuff. Mostly perf wise. No matter how you implement that it's going to be an indirection.

While you could say that we don't do runtime spell data access that much that's... not really true. Or at least my testing shows that it isn't.
I was looking at reducing simc binary size recently. This basically amounts to reducing the size of the different dbc structs. One thing that I already knew (and is a pet peeve of mine[*]) is that we use doubles *everywhere*, even for client originated data (which is 32b floats). A simple change from double to float for all dbc exported structs and I got a 4-5% slowdown depending on the profile... Not nice (but it reduced the binary size by ~5MB /shrug).

I did't really analyze the results (no `perf`) and I didn't really look how spell data is split up between the different db2 files. So maybe it wouldn't be as bad, but we must be wary of possible perf regressions with such a split up.

[*] So yeah, doubles. This is a huge pet peeve of mine. I think I remember asking Nate a long time ago why simc used doubles everywhere. I don't remember the exact answer, but I do remember it making sense then - smth along the lines of: we had no spell data back then, so it was all manual constants and using doubles was just easier (no requirement to sprinkle `f`s around).
But right now we mostly use client spell data, which is in floats. As I come from game dev myself I would be *very* surprised if there are actually any computations done with doubles in game (we build with `-Werror=double-promotion` for example). So we don't need the extended precision (I wouldn't even be surprised if some of the inconsistencies we get from the sim vs in-game numbers are due to us using doubles). But alas...

navv1234 commented:
> > My intent with the refactoring is to split everything into their own files instead of growing our primary structs even more (spell, item). For that to happen we need to figure out a way to do something like what you had (the pseudo-runtime linking of data X with our "spell_data_t" using spans or something similar).
> > That change will also mean that the whole interface on how we access that data could change, as there might be multiple entries in a linked db2 file for a single spell (might not be true for player-related spells though, this is more on the enemy side).
> 
> I'm not sure about splitting `spell_data_t` tbh. While it's natural for labels, powers and (especially) effects (it's always 1-to-n and the linked data is somewhat big) I'd be very cautious about splitting off other stuff. Mostly perf wise. No matter how you implement that it's going to be an indirection.

Yes this is the problem. I want to do it because in my opinion the spell data generator is too complicated right now. I want to simplify it. I also want to make changes more visible, figuring out what changed in our massive spell data from diffs is quite difficult in some cases. However, I am not willing to sacrifice performance too much there, if the indirection is a significant hit, then it's no go .. but there's no way of knowing that without doing some tests, and splitting up the huge struct is a pain.
 
> While you could say that we don't do runtime spell data access that much that's... not really true. Or at least my testing shows that it isn't.
> I was looking at reducing simc binary size recently. This basically amounts to reducing the size of the different dbc structs. One thing that I already knew (and is a pet peeve of mine[*]) is that we use doubles _everywhere_, even for client originated data (which is 32b floats). A simple change from double to float for all dbc exported structs and I got a 4-5% slowdown depending on the profile... Not nice (but it reduced the binary size by ~5MB /shrug).
> 
> I did't really analyze the results (no `perf`) and I didn't really look how spell data is split up between the different db2 files. So maybe it wouldn't be as bad, but we must be wary of possible perf regressions with such a split up.

Are we talking about spell data or effect access? I would expect spell data during iteration to be minimal (almost all modules cache pointers), however effect data is another thing entirely, that happens very often (`->effectN( x ).something()`). But that stuff is already behind an indirection.

Re: doubles, yes, I want to change client data export from double to float. It should have been like this from the start, but I suspect my original motivation when implementing the system was "simc uses doubles, so client data export uses doubles". And yes, client data export came later, if memory servers right around WotLK. If it's a noticeable performance regression though, I'm not sure we should do it. I'd rather take larger binary.

> 
> [*] So yeah, doubles. This is a huge pet peeve of mine. I think I remember asking Nate a long time ago why simc used doubles everywhere. I don't remember the exact answer, but I do remember it making sense then - smth along the lines of: we had no spell data back then, so it was all manual constants and using doubles was just easier (no requirement to sprinkle `f`s around).
> But right now we mostly use client spell data, which is in floats. As I come from game dev myself I would be _very_ surprised if there are actually any computations done with doubles in game (we build with `-Werror=double-promotion` for example). So we don't need the extended precision (I wouldn't even be surprised if some of the inconsistencies we get from the sim vs in-game numbers are due to us using doubles). But alas...

nuoHep commented:
> [ ... ]
> 
> Are we talking about spell data or effect access? I would expect spell data during iteration to be minimal (almost all modules cache pointers), however effect data is another thing entirely, that happens very often (`->effectN( x ).something()`). But that stuff is already behind an indirection.
> 

Spell data. What I mean is: say we split off `SpellAuraOptions.db2` data from `spell_data_t` into its own struct and slap a pointer into spell data to "link" it. While the `spell_data_t` pointers are cached, accessing any field from aura options now is behind another pointer. Aura options contains `proc_chance` field which, from a quick grep, is accessed quite a lot from `action_t::execute()`s in different modules.
It obviously has to be profiled.

And yeah, effects right now are behind a triple indirection (ptr to vector -> ptr to buffer -> ptr to effect). Reducing it was one of the motivations for my "spanification". It removed 1 indirection. And I think we can actually do even better (at the cost of slightly more complexity for runtime data overrides machinery).

> Re: doubles, yes, I want to change client data export from double to float. It should have been like this from the start, but I suspect my original motivation when implementing the system was "simc uses doubles, so client data export uses doubles". And yes, client data export came later, if memory servers right around WotLK. If it's a noticeable performance regression though, I'm not sure we should do it. I'd rather take larger binary.
> 
> [ ... ]

Thinking about it a bit more, what I think we can safely do is change all item-related exports to floats and leave spell/effect data as doubles (at least as a first step). While spell data gets a lot of runtime accesses (where an extra `cvtss2sd` actually slows down things) item data is accessed only on init afaik.

navv1234 commented:
> > [ ... ]
> > Are we talking about spell data or effect access? I would expect spell data during iteration to be minimal (almost all modules cache pointers), however effect data is another thing entirely, that happens very often (`->effectN( x ).something()`). But that stuff is already behind an indirection.
> 
> Spell data. What I mean is: say we split off `SpellAuraOptions.db2` data from `spell_data_t` into its own struct and slap a pointer into spell data to "link" it. While the `spell_data_t` pointers are cached, accessing any field from aura options now is behind another pointer. Aura options contains `proc_chance` field which, from a quick grep, is accessed quite a lot from `action_t::execute()`s in different modules.
> It obviously has to be profiled.

Right, yeah splitting off things would have implications on that for fields that are accessed during runtime.

> And yeah, effects right now are behind a triple indirection (ptr to vector -> ptr to buffer -> ptr to effect). Reducing it was one of the motivations for my "spanification". It removed 1 indirection. And I think we can actually do even better (at the cost of slightly more complexity for runtime data overrides machinery).
> 
> > Re: doubles, yes, I want to change client data export from double to float. It should have been like this from the start, but I suspect my original motivation when implementing the system was "simc uses doubles, so client data export uses doubles". And yes, client data export came later, if memory servers right around WotLK. If it's a noticeable performance regression though, I'm not sure we should do it. I'd rather take larger binary.
> > [ ... ]
> 
> Thinking about it a bit more, what I think we can safely do is change all item-related exports to floats and leave spell/effect data as doubles (at least as a first step). While spell data gets a lot of runtime accesses (where an extra `cvtss2sd` actually slows down things) item data is accessed only on init afaik.

Item data is only accessed during initialization, yes. I also want to split the item struct a bit (carve off item effects; we already have item effect export for the new item bonus system, so we should be able to do this for item effects easy enough.

nuoHep commented:
Closing. Rebased and resubmitted as #5180
